SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
+}
+
+// occupied with respect to vacuum reference!! defining the occupied orbitals in the Fermi vacuum is absolutely essential.
+IndexSpace vacuum_occupied()const {
+  if(vacuum_occupied_ == nulltype){throw std::invalid_argument("occupied has not been registered. please assign the occupied"
+"space label by calling assign_vacuum_occupied(label)");
+}
+else return vacuum_occupied_;
+}
+
+// space where active particles can reside, in Single Reference case this is usually active_occupied. Multireference
+// will also include an active space.
+IndexSpace active_particle_space()const {
+  if(active_particle_space_ == nulltype){throw std::invalid_argument("active_particle_space has not been registered. please assign the"
+"space label by calling assign_active_particle_space(label)");
+}
+else return active_particle_space_;
+}
+
+// space where all active holes can reside, in Single Reference case this is usually active_unoccupied. Multireference
+// will also include an active space.
+IndexSpace active_hole_space()const {
+  if(active_hole_space_ == nulltype){throw std::invalid_argument("active_hole_space has not been registered. please assign the"
+"space label by calling assign_hole_particle_space(label)");
+}
+else return active_hole_space_;
+}
+
+// needed to compute densites in physical vacuum.
+IndexSpace density_occupied()const {
+if(density_occuiped_ == nulltype_()){
+throw std::invalid_argument("density_occupied has not been registered. please assign the"
+"space label by calling assign_density_occupied(label)");
+}
+else{
+return density_occuiped_;
+}
+}
+void assign_vacuum_occupied(std::wstring label){
+  if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+vacuum_occupied_ = label_space[label];
+}
+
+void assign_complete(std::wstring label){
+if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+complete_ = label_space[label];
+}
+
+void assign_active_particle_space(std::wstring label){
+  if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+active_particle_space_ = label_space[label];
+}
+
+void assign_active_hole_space(std::wstring label){
+  if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+active_hole_space_ = label_space[label];
+}
+
+// which spaces could contain particles in your representation
+void assign_density_occupied(std::wstring label){
+if(label_space.find(label) == label_space.end()){
+throw std::invalid_argument("label not added to registry");
+}
+density_occuiped_ = label_space[label];
+}
+
+
+ private:
+  std::map<std::wstring ,IndexSpace> label_space;
+  const IndexSpace nulltype= {L"",0,0};
+
+  ///TODO use c++20 std::has_single_bit() when we update to this version
+  bool has_single_bit(std::uint32_t bits){
+    return bits&(((bool)(bits&(bits-1)))-1);
```

Comment:
```suggestion
    return bits && !(bits & (bits - 1));
```
conversion from integral types to booleans happens implicit